File: PaymentGateway/Python/pg_request.py
import asyncio
import random
import requests
import base64
import hashlib
from xml.etree import ElementTree

class pg_request:

    def __init__(self):
        self._http_client = requests.Session()

    async def make_get_query_transaction_id_request_async(self):
        transaction_id = 900497
        message_xml = f"""<?xml version="1.0" encoding="UTF-8"?><query><transactionid>{transaction_id}</transactionid></query>"""

        encoded_message = base64.b64encode(message_xml.encode('utf-8')).decode('utf-8')
        secret = "27f18bfcbe4d7f39971cb3460fbe7234a82fb48f985cf22a068fa1a685fe7e6f93c7d0d92fee4e8fd7dc0c9f11e2507300e675220ee85679afa681407ee2416d"
        mac = self.get_sha512_hash(encoded_message + secret)

        #print("mac:", mac)
        #print("encodedMessage:", encoded_message)

        try:
            url = "https://webpaypaymentgatewaystage.svea.com/webpay/rest/querytransactionid"
            content = {
                "merchantid": "1200",
                "message": encoded_message,
                "mac": mac
            }
            response = self._http_client.post(url, data=content)

            #print("Response status code:", response.status_code)
            #print("Response message:", response.text)

            if response.status_code == 200:
                print("Success!")
            else:
                print("Failed...")

            # Parsing the XML to find the 'message' element
            root = ElementTree.fromstring(response.text)
            message_element = root.find('.//message')
            if message_element is not None:
                encoded_message_part = message_element.text

                # Decoding the Base64 message part
                decoded_bytes = base64.b64decode(encoded_message_part)
                decoded_string = decoded_bytes.decode('utf-8')
                #print("Decoded message:")
                #print(decoded_string)
            else:
                print("Message element not found in response")
        except Exception as e:
            print(e)

    async def make_post_request_async(self):
        random_ref_no = random.randint(1000000, 9999999)
        message_xml = f"""<?xml version="1.0" encoding="UTF-8"?><payment><paymentmethod>SVEACARDPAY</paymentmethod><currency>SEK</currency><amount>500</amount><vat>100</vat><customerrefno>{random_ref_no}</customerrefno><returnurl>https://webpaypaymentgatewaystage.svea.com/webpay-admin/admin/merchantresponsetest.xhtml</returnurl><lang>en</lang></payment>"""

        encoded_message = base64.b64encode(message_xml.encode('utf-8')).decode('utf-8')
        secret = "27f18bfcbe4d7f39971cb3460fbe7234a82fb48f985cf22a068fa1a685fe7e6f93c7d0d92fee4e8fd7dc0c9f11e2507300e675220ee85679afa681407ee2416d"
        mac = self.get_sha512_hash(encoded_message + secret)

        #print("mac:", mac)
        #print("encodedMessage:", encoded_message)

        try:
            url = "https://webpaypaymentgatewaystage.svea.com/webpay/payment"
            content = {
                "merchantid": "1200",
                "message": encoded_message,
                "mac": mac
            }
            response = self._http_client.post(url, data=content)

            #print("Response status code:", response.status_code)
            #print("Response message:", response.text)
            # Since we receive an iframe and the status tends to return 400 - verify the HTML content instead
            if "enter your card details" in response.text.lower() or "select card type" in response.text.lower():
                print("Success!")
            else:
                print("Failed...")
        except Exception as e:
            print(e)

    def get_sha512_hash(self, input_str):
        hash_obj = hashlib.sha512()
        hash_obj.update(input_str.encode('utf-8'))
        return hash_obj.hexdigest()

async def main():
    tester = pg_request()
    print("Running GET request for PaymentGateway (Python)")
    await tester.make_get_query_transaction_id_request_async()
    print("Running Create request for PaymentGateway (Python)")
    await tester.make_post_request_async()
    print("----------------------------------------------------------")

if __name__ == "__main__":
    asyncio.run(main())

File: PaymentGateway/Php/pg_request.php
<?php

class PgRequest
{
    private $httpClient;

    public function __construct()
    {
        $this->httpClient = curl_init();
    }

    public function makeGetQueryTransactionIdRequestAsync()
    {
        $transactionId = PG_ORDER_TO_FETCH;
        $messageXML = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><query><transactionid>$transactionId</transactionid></query>";
        $encodedMessage = base64_encode($messageXML);
        $secret = "PG_SECRET_KEY";
        $mac = $this->getSha512Hash($encodedMessage . $secret);

        //echo "mac: " . $mac . "\n";
        //echo "encodedMessage: " . $encodedMessage . "\n";

        try {
            $url = "https://webpaypaymentgatewaystage.svea.com/webpay/rest/querytransactionid";
            $content = array(
                "merchantid" => "PG_MERCHANT_ID",
                "message" => $encodedMessage,
                "mac" => $mac
            );

            $response = $this->makeHttpRequest($url, $content);
            //echo "Response status code: " . $response['status_code'] . "\n";
            //echo "Response message: " . $response['body'] . "\n";

            $xml = simplexml_load_string($response['body']);
            $messageElement = $xml->xpath('//message');
            if ($messageElement) {
                $encodedMessagePart = (string)$messageElement[0];
                // Decoding the Base64 message part
                $decodedBytes = base64_decode($encodedMessagePart);
                $decodedString = utf8_decode($decodedBytes);
                //echo "Decoded message:\n";
                //echo $decodedString . "\n";
            } else {
                echo "Message element not found in response\n";
            }
            if ($response['status_code'] == 200)
                echo "Success!\n";
            else
                echo "Failed...\n";
        } catch (Exception $e) {
            echo $e->getMessage();
        }
    }

    public function makePostRequestAsync()
    {
        $randomRefNo = rand(1000000, 9999999);
        $messageXML = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><payment><paymentmethod>SVEACARDPAY</paymentmethod><currency>SEK</currency><amount>500</amount><vat>100</vat><customerrefno>$randomRefNo</customerrefno><returnurl>https://webpaypaymentgatewaystage.svea.com/webpay-admin/admin/merchantresponsetest.xhtml</returnurl><lang>en</lang></payment>";
        $encodedMessage = base64_encode($messageXML);
        $secret = "PG_SECRET_KEY";
        $mac = $this->getSha512Hash($encodedMessage . $secret);

        //echo "mac: " . $mac . "\n";
        //echo "encodedMessage: " . $encodedMessage . "\n";

        try {
            $url = "https://webpaypaymentgatewaystage.svea.com/webpay/payment";
            $content = array(
                "merchantid" => "PG_MERCHANT_ID",
                "message" => $encodedMessage,
                "mac" => $mac
            );

            $response = $this->makeHttpRequest($url, $content);
            //echo "Response status code: " . $response['status_code'] . "\n";
            $responseContent = $response['body'];
            //echo "Response message: " . $responseContent . "\n";
            // Since we receive an iframe and the status tends to return 400 - verify the HTML content instead
            if (strpos(strtolower($responseContent), "enter your card details") !== false || strpos(strtolower($responseContent), "select card type") !== false)
                echo "Success!\n";
            else
                echo "Failed...\n";

        } catch (Exception $e) {
            echo $e->getMessage();
        }
    }

    private function getSha512Hash($input)
    {
        return hash('sha512', $input);
    }

    private function makeHttpRequest($url, $postData)
    {
        curl_setopt_array($this->httpClient, array(
            CURLOPT_URL => $url,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => http_build_query($postData),
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HEADER => true,
            CURLOPT_FOLLOWLOCATION => true, // Follow redirects
            CURLOPT_MAXREDIRS => 10 // Maximum number of redirects to follow
        ));

        $response = curl_exec($this->httpClient);
        if (curl_errno($this->httpClient)) {
            throw new Exception(curl_error($this->httpClient));
        }

        $headerSize = curl_getinfo($this->httpClient, CURLINFO_HEADER_SIZE);
        $header = substr($response, 0, $headerSize);
        $body = substr($response, $headerSize);
        $statusCode = curl_getinfo($this->httpClient, CURLINFO_HTTP_CODE);

        return array(
            'status_code' => $statusCode,
            'header' => $header,
            'body' => $body
        );
    }
}

$tester = new PgRequest();
echo "Running Create request for PaymentGateway (Php)\n";
$tester->makePostRequestAsync();
echo "Running GET request for PaymentGateway (Php)\n";
$tester->makeGetQueryTransactionIdRequestAsync();

?>
